lib/rubocop/cop/rspec/expect_output.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks for opportunities to use `expect { ... }.to output`.
+      #
+      # @example
+      #   # bad
+      #   $stdout = StringIO.new
+      #   my_app.print_report
+      #   $stdout = STDOUT
+      #   expect($stdout.string).to eq('Hello World')
+      #
+      #   # good
+      #   expect { my_app.print_report }.to output('Hello World').to_stdout
+      class ExpectOutput < Cop
+        MSG = 'Use `expect { ... }.to output(...).to_%<name>s` ' \
+              'instead of mutating $%<name>s'.freeze
+
+        def_node_matcher :example?, Examples::ALL.block_pattern
+        def_node_matcher :hook?, Hooks::ALL.block_pattern
+
+        def on_gvasgn(node)
+          return unless inside_example_scope?(node)
+
+          variable_name, = *node
+          name = variable_name[1..-1]
+          return unless name.eql?('stdout') || name.eql?('stderr')
+
+          add_offense(node, :name, format(MSG, name: name))
+        end
+
+        private
+
+        def inside_example_scope?(node)
+          return false if node.nil? || example_group?(node)
+          return true if example?(node)
+          return RuboCop::RSpec::Hook.new(node).example? if hook?(node)
```

Comment:
We only want to flag overwriting of `$stdout` or `$stderr` in the context of an example because we are looking for the pattern of

1. Write to `$stdout`
2. Run task being tested
3. Check value of stdout
4. Reset global variable

If someone is altering either global variable in `before(:all)` or `before(:suite)` then something else is going on that doesn't relate to this cop